File: modules/graphics/lib/aerox/graphics/shaders/Shader.cpp
#include "aerox/graphics/shaders/Shader.hpp"
#include "aerox/core/utils.hpp"
#include "aerox/graphics/descriptors/DescriptorLayoutBuilder.hpp"
#include "ashl/utils.hpp"

namespace aerox::graphics
{
    ShaderManager* Shader::GetManager() const
    {
        return _shaderManager;
    }

    Shader::PushConstant::PushConstant(const std::string& inName, const uint64_t& inSize,
                                       const vk::ShaderStageFlags& inStages)
    {
        name = inName;
        size = inSize;
        stages = inStages;
    }

    void Shader::ComputeResources(
        const std::vector<std::pair<vk::ShaderStageFlags, std::shared_ptr<ashl::ModuleNode>>>& shaders)
    {
        for (auto& [shaderStages,shader] : shaders)
        {
            for (auto& node : shader->statements)
            {
                switch (node->nodeType)
                {
                case ashl::NodeType::Layout:
                    {
                        if (auto asLayout = std::dynamic_pointer_cast<ashl::LayoutNode>(node); asLayout && asLayout->
                            layoutType == ashl::ELayoutType::Uniform)
                        {
                            vk::DescriptorType type{};
                            auto stages = shaderStages;
                            switch (asLayout->declaration->declarationType)
                            {
                            case ashl::EDeclarationType::Block:
                                type = asLayout->tags.contains("$storage")
                                           ? vk::DescriptorType::eStorageBuffer
                                           : vk::DescriptorType::eUniformBuffer;
                                break;
                            case ashl::EDeclarationType::Sampler2D:
                                type = vk::DescriptorType::eCombinedImageSampler;
                                break;
                            case ashl::EDeclarationType::Sampler:
                                type = vk::DescriptorType::eSampler;
                                break;
                            case ashl::EDeclarationType::Texture2D:
                                type = vk::DescriptorType::eSampledImage;
                                break;
                            default:
                                throw std::runtime_error("Unexpected declaration type");
                            }

                            auto set = static_cast<unsigned>(
                                ashl::parseInt(asLayout->tags["set"]));
                            auto binding = static_cast<unsigned>(ashl::parseInt(
                                asLayout->tags["binding"]));

                            vk::DescriptorBindingFlags bindingFlags{};


                            if(asLayout->tags.contains("$update"))
                            {
                                bindingFlags |= vk::DescriptorBindingFlagBits::eUpdateAfterBind;
                            }
                            
                            if (asLayout->tags.contains("$partial"))
                            {
                                bindingFlags |= vk::DescriptorBindingFlagBits::ePartiallyBound;
                            }
                            
                            if (asLayout->tags.contains("$variable"))
                            {
                                bindingFlags |= vk::DescriptorBindingFlagBits::eVariableDescriptorCount;
                            }

                            if (asLayout->tags.contains("$stage") && asLayout->tags["$stage"] == "all")
                            {
                                stages |= vk::ShaderStageFlagBits::eAll;
                            }

                            auto name = asLayout->declaration->declarationName;

                            if (resources.contains(name))
                            {
                                resources.at(name).stages |= stages;
                                resources.at(name).bindingFlags |= bindingFlags;
                            }
                            else
                            {
                                auto count = static_cast<unsigned>(asLayout->declaration->
                                                              declarationCount);
                                if(asLayout->tags.contains("$variable"))
                                {
                                    count = ashl::parseInt(asLayout->tags.at("$variable"));
                                }
                                
                                if (asLayout->declaration->declarationType == ashl::EDeclarationType::Block)
                                {
                                    resources.emplace(name, ShaderResource{
                                                          name,
                                                          set,
                                                          binding,
                                                          count,
                                                          type,
                                                          stages,
                                                          bindingFlags,
                                                          static_cast<uint32_t>(asLayout->declaration->GetSize())
                                                      });
                                }
                                else
                                {
                                    resources.emplace(name, ShaderResource{
                                                          name,
                                                          set,
                                                          binding,
                                                          count,
                                                          type,
                                                          stages,
                                                          bindingFlags
                                                      });
                                }
                            }
                        }
                    }
                    break;
                case ashl::NodeType::PushConstant:
                    {
                        if (auto asPushConstant = std::dynamic_pointer_cast<ashl::PushConstantNode>(node))
                        {
                            std::string name = "push";
                            if (pushConstants.contains(name))
                            {
                                pushConstants.at(name).stages |= shaderStages;
                            }
                            else
                            {
                                uint64_t pushSize = 0;
                                for (auto& declarationNode : asPushConstant->declarations)
                                {
                                    pushSize += declarationNode->GetSize();
                                }

                                pushConstants.emplace(name, PushConstant{name, pushSize, shaderStages});
                            }
                        }
                    }
                    break;
                default:
                    break;
                }
            }
        }
    }

    vk::ShaderStageFlags Shader::ScopeTypeToStageFlags(const ashl::EScopeType& scopeType)
    {
        switch (scopeType)
        {
        case ashl::EScopeType::Fragment:
            return vk::ShaderStageFlagBits::eFragment;
        case ashl::EScopeType::Vertex:
            return vk::ShaderStageFlagBits::eVertex;
        }

        return vk::ShaderStageFlagBits::eVertex;
    }

    Shader::Shader(ShaderManager* manager)
    {
        _shaderManager = manager;
    }

    std::map<uint32_t, vk::DescriptorSetLayout> Shader::ComputeDescriptorSetLayouts() const
    {
        
        std::map<uint32_t, DescriptorLayoutBuilder> builders{};

        for (auto& item : resources | std::views::values)
        {
            if (!builders.contains(item.set))
                builders.emplace(item.set,
                                 DescriptorLayoutBuilder{});

            builders.at(item.set).AddBinding(item.binding,item.type,item.stages,item.count,item.bindingFlags);
        }
        std::map<uint32_t,vk::DescriptorSetLayout> layouts{};
        for (auto &[set,builder] : builders)
        {
            layouts.emplace(set,builder.Build());
        }
        return layouts;
    }

    std::vector<vk::PushConstantRange> Shader::ComputePushConstantRanges() const
    {
        std::vector<vk::PushConstantRange> pushConstantRanges{};
        for (auto& item : pushConstants | std::views::values)
        {
            pushConstantRanges.emplace_back(item.stages, 0, static_cast<uint32_t>(item.size));
        }
        return pushConstantRanges;
    }
}

File: aerox.Runtime.Graphics/Shaders/Shader.cs
using aerox.Runtime.Graphics.Descriptors;
using rsl.Nodes;
using TerraFX.Interop.Vulkan;

namespace aerox.Runtime.Graphics.Shaders;

public abstract class Shader : Disposable
{
    
    public class Resource
    {
        public string Name = "";
        public uint Set;
        public uint Binding;
        public uint Count;
        public VkDescriptorType Type;
        public VkShaderStageFlags Stages;
        public VkDescriptorBindingFlags BindingFlags;
        public uint Size = 0;
    }
    
    public class PushConstant
    {
        public string Name = "";
        public ulong Size;
        public VkShaderStageFlags Stages;
    }

    public Dictionary<string, Resource> Resources = [];
    public Dictionary<string, PushConstant> PushConstants = [];
    
    public abstract bool Bind(VkCommandBuffer cmd,bool wait = false);

    public abstract CompiledShader Compile(ShaderManager manager);

    protected void ComputeResources(Pair<VkShaderStageFlags, ModuleNode>[] shaders)
    {
        foreach (var (shaderStages,shader) in shaders)
        {
            foreach (var node in shader.Statements)
            {
                switch (node.NodeType)
                {
                    case NodeType.Layout:
                    {
                        if (node is LayoutNode asLayout && asLayout.LayoutType == LayoutType.Uniform)
                        {
                            VkDescriptorType type = asLayout.Declaration.DeclarationType switch
                            {
                                DeclarationType.Block => asLayout.Tags.ContainsKey("$storage") ? VkDescriptorType.VK_DESCRIPTOR_TYPE_STORAGE_BUFFER : VkDescriptorType.VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER,
                                DeclarationType.Sampler2D => VkDescriptorType.VK_DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER,
                                _ => throw new Exception("Unexpected declaration type")
                            };

                            var set = uint.Parse(asLayout.Tags["set"]);
                            var binding = uint.Parse(asLayout.Tags["binding"]);

                            VkDescriptorBindingFlags bindingFlags = 0;
                            VkShaderStageFlags stages = 0;

                            if (asLayout.Tags.ContainsKey("$update"))
                            {
                                bindingFlags |= VkDescriptorBindingFlags.VK_DESCRIPTOR_BINDING_UPDATE_AFTER_BIND_BIT;
                            }
                            
                            if (asLayout.Tags.ContainsKey("$partial"))
                            {
                                bindingFlags |= VkDescriptorBindingFlags.VK_DESCRIPTOR_BINDING_PARTIALLY_BOUND_BIT;
                            }
                            
                            if (asLayout.Tags.ContainsKey("$variable"))
                            {
                                bindingFlags |= VkDescriptorBindingFlags.VK_DESCRIPTOR_BINDING_VARIABLE_DESCRIPTOR_COUNT_BIT;
                            }
                            
                            if (asLayout.Tags.ContainsKey("$stage") && asLayout.Tags["$stage"] == "all")
                            {
                                stages |= VkShaderStageFlags.VK_SHADER_STAGE_ALL;
                            }

                            var name = asLayout.Declaration.DeclarationName;

                            if (Resources.ContainsKey(name))
                            {
                                Resources[name].Stages |= stages;
                                Resources[name].BindingFlags |= bindingFlags;
                            }
                            else
                            {
                                var count = (uint)asLayout.Declaration.Count;

                                if (asLayout.Tags.ContainsKey("$variable"))
                                {
                                    count = uint.Parse(asLayout.Tags["$variable"]);
                                }

                                if (asLayout.Declaration.DeclarationType == DeclarationType.Block)
                                {
                                    Resources.Add(name,new Resource()
                                    {
                                        Name = name,
                                        Set = set,
                                        Binding = binding,
                                        Count = count,
                                        Type = type,
                                        Stages = stages,
                                        BindingFlags = bindingFlags,
                                        Size = (uint)asLayout.Declaration.SizeOf()
                                    });
                                }
                                else
                                {
                                    Resources.Add(name,new Resource()
                                    {
                                        Name = name,
                                        Set = set,
                                        Binding = binding,
                                        Count = count,
                                        Type = type,
                                        Stages = stages,
                                        BindingFlags = bindingFlags
                                    });
                                }
                            }
                        }
                    }
                        break;
                    case NodeType.PushConstant:
                    {
                        if (node is PushConstantNode asPushConstant)
                        {
                            var name = "push";
                            if (PushConstants.TryGetValue(name, out var constant))
                            {
                                constant.Stages |= shaderStages;
                            }
                            else
                            {
                                ulong pushSize = 0;
                                foreach (var declarationNode in asPushConstant.Declarations)
                                {
                                    pushSize += (ulong)declarationNode.SizeOf();
                                }
                                
                                PushConstants.Add(name,new PushConstant()
                                {
                                    Name = name,
                                    Size = pushSize,
                                    Stages = shaderStages
                                });
                            }
                        }
                    }
                        break;
                    default:
                        break;
                }
            }
        }
    }

    public abstract void Init();
    public abstract Dictionary<uint, VkDescriptorSetLayout> GetDescriptorSetLayouts();
    public abstract VkPipelineLayout GetPipelineLayouts();
    protected override void OnDispose(bool isManual)
    {
        throw new NotImplementedException();
    }
}